File: src/main/java/com/kasisoft/libs/common/types/HasFirstAndLast.java
package com.kasisoft.libs.common.types;

import jakarta.validation.constraints.*;

import java.util.*;

/**
 * @author daniel.kasmeroglu@kasisoft.com
 */
public interface HasFirstAndLast<F, L> {

    @NotNull
    Optional<F> findFirst();

    @NotNull
    Optional<L> findLast();

    default F getFirst() {
        return findFirst().orElse(null);
    }

    default L getLast() {
        return findLast().orElse(null);
    }

} /* ENDINTERFACE */

File: src/main/kotlin/com/kasisoft/libs/common/types/HasFirstAndLast.kt
package com.kasisoft.libs.common.types

import com.kasisoft.libs.common.*

import java.util.*

interface HasFirst<T> {

    fun findFirst(): T?

    fun getFirst(): T = findFirst() ?: throw KclException("no first element")

} /* ENDINTERFACE */

interface HasLast<T> {

    fun findLast(): T?

    fun getLast(): T = findLast() ?: throw KclException("no last element")

} /* ENDINTERFACE */

interface HasFirstAndLast<F, L>: HasFirst<F>, HasLast<L>
